Raise a more specific error when the job class can't be instantiated

Earlopain opened the pull request:
### Motivation / Background

During a rolling deploy, jobs maybe be enqueued that the job processor doesn't know about yet. Currently it's not easily possible for an adapter to know if the exception originated from user code or not.

The same can happen in the other direction of course (job got enqueued and new codebase doesn't contain it anymore), this just allows adapters to decide on what happens.

### Detail

Add a new exception to raise that inherits from `NameError` for compatibility.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

byroot commented on activejob/lib/active_job/core.rb:
```diff
@@ -1,6 +1,13 @@
 
 module ActiveJob
+  # Raised during job payload deserialization when it references an uninitialized job class.
+  class UninitializedJobClassError < NameError
```
```suggestion
  class UnknownJobClassError < NameError
```

Sounds like a better name?

byroot commented on activejob/lib/active_job/core.rb:
```diff
@@ -1,6 +1,13 @@
+  # Raised during job payload deserialization when it references an uninitialized job class.
+  class UninitializedJobClassError < NameError
+    def initialize(job_class_name)
+      super("Failed to instantiate job class #{job_class_name}", job_class_name)
```
Might as well explain the class doesn't exist?

Earlopain committed:
Raise a more specific error when the job class can't be instantiated

During a rolling deploy, jobs maybe be enqueued that the job processor
doesn't know about yet. Currently it's not easily possible for the job
adapter to know if the exception originated from user code or not.

Earlopain commented:
Thank you, that was quick!
